simulation/halsim_gui/src/main/native/cpp/DriverStationGui.cpp

```diff
@@ -1100,38 +1105,47 @@ static void DriverStationExecute() {
 
     ImGui::SetNextWindowPos(ImVec2{5, 20}, ImGuiCond_FirstUseEver);
     ImGui::Begin("Robot State", nullptr, ImGuiWindowFlags_AlwaysAutoResize);
-    if (ImGui::Selectable("Disabled", !isEnabled) || disableHotkey) {
+    if (ImGui::Selectable("Disconnected", !isAttached)) {
       HALSIM_SetDriverStationEnabled(false);
+      HALSIM_SetDriverStationDsAttached(false);
+      isAttached = false;
+      gFMSModel->Update();
     }
-    if (ImGui::Selectable("Autonomous", isEnabled && isAuto && !isTest)) {
-      HALSIM_SetDriverStationAutonomous(true);
-      HALSIM_SetDriverStationTest(false);
-      HALSIM_SetDriverStationEnabled(true);
+    if (ImGui::Selectable("Disabled", isAttached && !isEnabled) ||
+        disableHotkey) {
+      DriverStationConnect(false, false, false);
     }
-    if (ImGui::Selectable("Teleoperated", isEnabled && !isAuto && !isTest) ||
+    if (ImGui::Selectable("Autonomous",
+                          isAttached && isEnabled && isAuto && !isTest)) {
+      DriverStationConnect(true, true, false);
+    }
+    if (ImGui::Selectable("Teleoperated",
+                          isAttached && isEnabled && !isAuto && !isTest) ||
         enableHotkey) {
-      HALSIM_SetDriverStationAutonomous(false);
-      HALSIM_SetDriverStationTest(false);
-      HALSIM_SetDriverStationEnabled(true);
+      DriverStationConnect(true, false, false);
     }
     if (ImGui::Selectable("Test", isEnabled && isTest)) {
-      HALSIM_SetDriverStationAutonomous(false);
-      HALSIM_SetDriverStationTest(true);
-      HALSIM_SetDriverStationEnabled(true);
+      DriverStationConnect(true, false, true);
     }
     ImGui::End();
   }
 
   // Update HAL
-  for (int i = 0, end = gRobotJoysticks.size();
-       i < end && i < HAL_kMaxJoysticks; ++i) {
-    gRobotJoysticks[i].SetHAL(i);
+  if (isAttached) {
+    for (int i = 0, end = gRobotJoysticks.size();
+         i < end && i < HAL_kMaxJoysticks; ++i) {
+      gRobotJoysticks[i].SetHAL(i);
+    }
   }
 
   // Send new data every 20 ms (may be slower depending on GUI refresh rate)
   static double lastNewDataTime = 0.0;
-  if ((curTime - lastNewDataTime) > 0.02 && !HALSIM_IsTimingPaused()) {
+  if ((curTime - lastNewDataTime) > 0.02 && !HALSIM_IsTimingPaused() &&
+      isAttached) {
```

Comment:
Is `HALSIM_IsTimingPaused()` expensive to call? If so, the `isAttached` should be moved earlier to avoid unnecessarily calling it (but if it's cheap then it doesn't really matter).

Reply:
It’s extremely cheap, just an atomic read.